### PR TITLE
[bugfix] Fix dependency bug in async policy

### LINKED ISSUE #1082: Async policy enters infinite loop when executing dependencies
Try inserting a failure before/after the setup phase in a test with dependencies:

```diff
--- a/unittests/resources/checks_unlisted/deps_complex.py
+++ b/unittests/resources/checks_unlisted/deps_complex.py
@@ -138,6 +138,10 @@ class T6(BaseTest):
         with open(os.path.join(T5().stagedir, 'out.txt')) as fp:
             self._count += int(fp.read())

+    @rfm.run_after('setup')
+    def fail(self):
+        raise Exception
+

 @rfm.simple_test
 class T7(BaseTest):
```

Then run as follows:

```
./bin/reframe -c unittests/resources/checks_unlisted/deps_complex.py --exec-policy=async -r
```

Output:

```
[==========] Running 8 check(s)
[==========] Started on Tue Dec  3 23:31:56 2019

[----------] started processing T0 (T0)
[ RUN      ] T0 on generic:login using builtin-gcc
[----------] finished processing T0 (T0)

[----------] started processing T4 (T4)
[ RUN      ] T4 on generic:login using builtin-gcc
[      DEP ] T4 on generic:login using builtin-gcc
[----------] finished processing T4 (T4)

[----------] started processing T5 (T5)
[ RUN      ] T5 on generic:login using builtin-gcc
[      DEP ] T5 on generic:login using builtin-gcc
[----------] finished processing T5 (T5)

[----------] started processing T1 (T1)
[ RUN      ] T1 on generic:login using builtin-gcc
[      DEP ] T1 on generic:login using builtin-gcc
[----------] finished processing T1 (T1)

[----------] started processing T6 (T6)
[ RUN      ] T6 on generic:login using builtin-gcc
[      DEP ] T6 on generic:login using builtin-gcc
[----------] finished processing T6 (T6)

[----------] started processing T2 (T2)
[ RUN      ] T2 on generic:login using builtin-gcc
[      DEP ] T2 on generic:login using builtin-gcc
[----------] finished processing T2 (T2)

[----------] started processing T3 (T3)
[ RUN      ] T3 on generic:login using builtin-gcc
[      DEP ] T3 on generic:login using builtin-gcc
[----------] finished processing T3 (T3)

[----------] started processing T7 (T7)
[ RUN      ] T7 on generic:login using builtin-gcc
[      DEP ] T7 on generic:login using builtin-gcc
[----------] finished processing T7 (T7)

[----------] waiting for spawned checks to finish
[       OK ] T0 on generic:login using builtin-gcc
[       OK ] T4 on generic:login using builtin-gcc
[       OK ] T5 on generic:login using builtin-gcc
[       OK ] T1 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
[     FAIL ] T6 on generic:login using builtin-gcc
```

### PR DESCRIPTION
**Pull request by ekouts**
Async policy enters infinite loop when executing dependencies that fail before/after the setup phase.

I also made the dependency unittests more strict so that they check that all cases either were successful or failed. Checking the `num_cases` of `stats` is not enough to make sure a testcase was accounted for in the final summary.

Fixes #1082 .

### CODE REVIEW

File: unittests/test_policies.py [hunk 1]
```diff
@@ -211,12 +226,13 @@ def test_dependencies(self):
         self.runall(self.checks, sort=True)
 
         stats = self.runner.stats
-        assert stats.num_cases(0) == 8
-        assert len(stats.failures()) == 2
+        assert stats.num_cases(0) == 10
+        self.assertRunall()
```
Comment (vkarak) [inline unittests/test_policies.py]: I would move this before the assignment to `stats`.

File: reframe/frontend/executors/policies.py [hunk 2]
```diff
@@ -295,22 +304,28 @@ def _poll_tasks(self):
...
+                               task.testcase.environ,
+                               sched_flex_alloc_nodes=self.sched_flex_alloc_nodes,
+                               sched_account=self.sched_account,
+                               sched_partition=self.sched_partition,
+                               sched_reservation=self.sched_reservation,
+                               sched_nodelist=self.sched_nodelist,
+                               sched_exclude_nodelist=self.sched_exclude_nodelist,
+                               sched_options=self.sched_options)
+                else:
+                    still_waiting.append(task)
```
[the remark this answers is not in the document]
Comment (ekouts) [reply reframe/frontend/executors/policies.py]: The logic inside the try of the two function is similar but also a bit different.
In `runcase()`:
1. if deps_failed() then fail
2. if not deps_succeeded then print a message + move to waiting
3. else (if deps_succeded) then setup + try to run it

Inside `_setup_all()`:
1. if deps_failed() then fail
2.  if deps_succeded then setup
3. else (if not deps_succeeded) keep the task in a still_waiting list

What would a more generic function do, without changing too much the logic of the functions?

[the remark this answers is not in the document]
Comment (vkarak) [reply reframe/frontend/executors/policies.py]: I see, what I don't like though is the extra `try/except`. Too much nesting. So my idea was to move this block in a function that you could potentially reuse, similarly to what we are doing with the `_reschedule()`. I don't have a problem in changing the logic or moving the printing of DEP.

File: reframe/frontend/executors/policies.py [hunk 3]
```diff
@@ -295,21 +320,11 @@ def _poll_tasks(self):
...
-                           sched_partition=self.sched_partition,
-                           sched_reservation=self.sched_reservation,
-                           sched_nodelist=self.sched_nodelist,
-                           sched_exclude_nodelist=self.sched_exclude_nodelist,
-                           sched_options=self.sched_options)
-            else:
+        # It needs to loop over a copy of _waiting_tasks because a testcase
+        # could fail on setup and be removed from the list _waiting_tasks.
+        # This would not be not safe and could have unxpected consequences.
+        for task in self._waiting_tasks[:]:
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: I don't like this copy. We are also doing another one later on, copying from the `still_waiting`. A more clever way could be to iterate the `waiting_tasks` from left to right, always popping the left item, which you should reinsert at the left if it should still  wait. You stop iterating until you  reach the end of the list. This way you can get rid of the `_remove_from_waiting()` and of the `still_waiting` here.

**Pull request merged**